Fix test runner when name contains a `/`

davidstosik opened the pull request:
## What this fixes

Here's the scenario I'd like to fix:
```rb
class MyTest < ActiveSupport::TestCase
  test "there is a / in here" do
    assert true
  end
end
```
```sh
bin/rails test test/my_test.rb --name test_there_is_a_/_in_here
```
```
`require': cannot load such file -- /tmp/testapp/test_there_is_a_/_in_here (LoadError)
```

## What the problem is

The main problem is that the code currently considers that any arg that contains at least one slash, no spaces, and is not surrounded with slashes has to be a file path, whereas in practice, arguments that follow `--name` and `-n` will never be a file path, _even_ if they contain a `/`.

## Reproducing

I've added some tests that prove the problem and fix, but here is another way to reproduce:

```
$ cd /tmp
$ rails new --minimal testapp
$ cd testapp
$ rails g scaffold Post title:string
```

Add a test to `test/models/post_test.rb` with a slash in its label:

```rb
test "dummy test with a / slash" do
  assert true
end
```

### Failure pattern 1

The following command should only run that single test, but instead it crashes on "cannot load such file":

```
$ bin/rails test test/models/post_test.rb -n test_dummy_test_with_a_/_slash
/opt/rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `require': cannot load such file -- /tmp/testapp/test_dummy_test_with_a_/_slash (LoadError)
```

`test_dummy_test_with_a_/_slash` was supposed to be a test name filter but was instead seen as a file path. Yet, the following command works:

```
$ bin/rails test test/modes/post_test.rb -n /test_dummy_test_with_a_\/_slash/
```

### Failure pattern 2

Another failure, which is probably less important, but happens to get fixed by this PR, is when passing an absolute file path to a test directory in order to run all tests defined in that directory.

Because the file path starts and ends with `/`, it is seen as a regular expression instead.

This next command should only run tests located in `test/models` but it runs everything:

```
$ bin/rails test --verbose /tmp/testapp/test/models/
Running 9 tests in a single process (parallelization threshold is 50)
Run options: --verbose --seed 32902

PostTest#test_the_truth = 0.02 s = .
PostTest#test_dummy_test_with_a_/_slash = 0.00 s = .
PostsControllerTest#test_should_show_post = 0.11 s = .
PostsControllerTest#test_should_get_index = 0.00 s = .
PostsControllerTest#test_should_get_edit = 0.01 s = .
PostsControllerTest#test_should_update_post = 0.01 s = .
PostsControllerTest#test_should_create_post = 0.01 s = .
PostsControllerTest#test_should_destroy_post = 0.00 s = .
PostsControllerTest#test_should_get_new = 0.00 s = .

Finished in 0.168346s, 53.4614 runs/s, 77.2221 assertions/s.
9 runs, 13 assertions, 0 failures, 0 errors, 0 skips
```

## Caveats

The code in `railties/lib/rails/test_unit/runner.rb` is hard to adjust. It's attempting to parse arguments and change them before eventually forwarding them to Minitest. Not using `OptionParser` means the code needs to somehow reinvent the wheel and handle multiple edge cases.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

davidstosik committed:
Fix test runner when name contains a /

 ## What this fixes

Here's the scenario I'd like to fix:
```rb
class MyTest < ActiveSupport::TestCase
  test "there is a / in here" do
    assert true
  end
end
```
```sh
bin/rails test test/my_test.rb --name test_there_is_a_/_in_here
```
```
`require': cannot load such file -- /tmp/testapp/dummy_test_with_a_/_slash (LoadError)
```

 ## What the problem is

The main problem is that the code currently considers that any arg that contains at least one slash, no spaces, and is not surrounded with slashes has to be a file path, whereas in practice, arguments that follow `--name` and `-n` will never be a file path, _even_ if they contain a `/`.

 ## Reproducing

I've added some tests that prove the problem and fix, but here is
another way to reproduce:

```
$ cd /tmp
$ rails new --minimal testapp
$ cd testapp
$ rails g scaffold Post title:string
```

Add a test to `test/models/post_test.rb` with a slash in its label:

```rb
test "dummy test with a / slash" do
  assert true
end
```

 ### Failure pattern 1

The following command should only run that single test, but instead it crashes on "cannot load such file":

```
$ bin/rails test test/models/post_test.rb -n test_dummy_test_with_a_/_slash
/opt/rubies/ruby-3.3.0/lib/ruby/3.3.0/bundled_gems.rb:74:in `require': cannot load such file -- /tmp/testapp/test_dummy_test_with_a_/_slash (LoadError)
```

`test_dummy_test_with_a_/_slash` was supposed to be a test name filter but was instead seen as a file path.
Yet, the following command works:

```
$ bin/rails test test/modes/post_test.rb -n /test_dummy_test_with_a_\/_slash/
```

 ### Failure pattern 2

Another failure, which is probably less important, but happens to get
fixed by this PR, is when passing an absolute file path to a test
directory in order to run all tests defined in that directory.

Because the file path starts and ends with `/`, it is seen as a regular
expression instead.

This next command should only run tests located in `test/models` but it runs everything:

```
$ bin/rails test --verbose /tmp/testapp/test/models/
Running 9 tests in a single process (parallelization threshold is 50)
Run options: --verbose --seed 32902

PostTest#test_the_truth = 0.02 s = .
PostTest#test_dummy_test_with_a_/_slash = 0.00 s = .
PostsControllerTest#test_should_show_post = 0.11 s = .
PostsControllerTest#test_should_get_index = 0.00 s = .
PostsControllerTest#test_should_get_edit = 0.01 s = .
PostsControllerTest#test_should_update_post = 0.01 s = .
PostsControllerTest#test_should_create_post = 0.01 s = .
PostsControllerTest#test_should_destroy_post = 0.00 s = .
PostsControllerTest#test_should_get_new = 0.00 s = .

Finished in 0.168346s, 53.4614 runs/s, 77.2221 assertions/s.
9 runs, 13 assertions, 0 failures, 0 errors, 0 skips
```

 ## Caveats

The code in `railties/lib/rails/test_unit/runner.rb` is hard to adjust.
It's attempting to parse arguments and change them before eventually forwarding them to Minitest.
Not using `OptionParser` means the code needs to somehow reinvent the
wheel and handle multiple edge cases.

davidstosik commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -82,9 +81,17 @@ def compose_filter(runnable, filter)
+                next
+              end
+
+              next if path.start_with?("-")
```
Not happy about this: a file path could possibly start with `-` (though that sounds unlikely).

rafaelfranca commented:
CI is broken. Can you take a look?

byroot commented:
Thanks for the patch, this bug has annoyed me for a long time, we're overdue to handle it.

Patch looks good, except the constant you removed broke a lot of tests.

davidstosik commented:
Sorry, I haven't got to it yet, will take a look soon. 🙇🏻‍♂️

davidstosik committed:
Fix constant that was using deleted PATH_ARGUMENT_PATTERN

davidstosik commented on railties/lib/rails/commands/test/test_command.rb:
```diff
@@ -69,7 +69,8 @@ def generators(*args)
 
       private
-        EXACT_TEST_ARGUMENT_PATTERN = /^-n|^--name\b|#{Rails::TestUnit::Runner::PATH_ARGUMENT_PATTERN}/
+        EXACT_TEST_ARGUMENT_PATTERN = /^(-n\b|--name\b|[^-])/
+        private_constant :EXACT_TEST_ARGUMENT_PATTERN
```
Two things:
1. `EXACT_TEST_ARGUMENT_PATTERN` is meant to match arguments which are any of those three:
    1. `-n`
    2. `--name`
    3. Any test file path.
      This one used to be based on `PATH_ARGUMENT_PATTERN` which was part of the problem this PR is solving.
      Realistically, any argument that does not start with a hyphen `-` _can be_ a test file path argument.
      In practice, it might not always be: it could be for example a seed passed to `--seed` or an exclude path passed to `--exclude` but since the code in Rails seems to [avoid using `OptionParser` to parse all Minitest options](https://github.com/rails/rails/blob/d444e7ff8766f67dbfe953d7eb9058e1cb0bf6c9/railties/lib/rails/test_unit/runner.rb#L32) (too tight coupling?), this seemed to be the easy way out.
      If there was a way to reuse Minitest's option parser, we could do better.
2. I might as well mark the constant as private, since it was meant to be.

davidstosik replied:
Turns out I was wrong, this is not an easy way out!
```
Error:
Rails::Command::TestTest#test_test_command_with_options_runs_test:prepare_task:
RuntimeError: rails command failed (1): bin/rails test --seed 1234 -e development 2>&1
/usr/local/lib/ruby/3.3.0/bundled_gems.rb:74:in `require': cannot load such file -- /rails/tmp/d20240623-1161-xi2hba/app/1234 (LoadError)
    from /usr/local/lib/ruby/3.3.0/bundled_gems.rb:74:in `block (2 levels) in replace_require'
```

davidstosik commented:
OK, so I've tried a few things, and none of them seem to work.

Basically, the code in Rails was based on the assumption that: a file path passed to the test command will contain at least one `/`, but not be surrounded by slashes, and any path that does not, and any argument that does not match these two condition is not one.

Note that `#extract_filters` will also return path patterns passed to `--exclude`, which means that Rails will try to require paths that were meant to be excluded. A simple example:
```
rails new testapp --minimal
cd testapp
bin/rails test --exclude test/does_not_exist.rb

(...):in `require': cannot load such file -- /private/tmp/testapp/test/does_not_exist.rb (LoadError)
```
Note that however, it works if the excluded path contains no `/` 😅 :
```
rails new testapp --minimal
cd testapp
bin/rails test --exclude testRunning 0 tests in a single process (parallelization threshold is 50)
Run options: --exclude test --seed 50505

# Running:

Finished in 0.000277s, 0.0000 runs/s, 0.0000 assertions/s.
0 runs, 0 assertions, 0 failures, 0 errors, 0 skips
```


Personally, since a proper, bug-free option parsing will have to rely on a lot of knowledge about what options can be passed to Minitest (which options take a parameter, like `--seed`, and which don't, like `--quiet`, which options take a path, that can potentially be altered by Rails, etc).

With this in mind, I think it would be more productive to either reimplement Minitest's `OptionParser` (defined [here](https://github.com/minitest/minitest/blob/0719fdbef0e5360369ffdcb05c2ee71056656ccf/lib/minitest.rb#L149-L187)), or directly use `Minitest.process_args` (defined [here](https://github.com/minitest/minitest/blob/0719fdbef0e5360369ffdcb05c2ee71056656ccf/lib/minitest.rb#L143C12-L143C24)), which sounds doable since that method will process options declared by the rails extension (defined [here](https://github.com/rails/rails/blob/da1ad718998ea700f9a1cda935a16d438c7b60ed/railties/lib/minitest/rails_plugin.rb#L71)).

rafaelfranca commented:
I don't think it is a problem to load the plugin early. We can load as soon as we want. If that would fix the issues you are having we should do it.

davidstosik commented:
Thanks @rafaelfranca, I'll take another look soon.

byroot commented:
Poke 😉

davidstosik commented:
Oof, sorry! 🙇🏻‍♂️ 
Parental leave got in the way and I forgot about it...
This is on my to-do list though, so I'll take a look soon. (🤞🏻)

davidstosik committed:
Fix test leak due to using class instance variables and inheritance

Here's a simple script to illustrate the problem:

```rb
class Foo
  def initialize
    @@options ||= "Foo @@options initialized!"
  end

  def get_options
    @@options
  end
end

class Bar < Foo
  def reset_options
    @@options = "@@options reset in Bar!"
  end
end

Bar.new.reset_options

puts Foo.new.get_options
 #=> "@@options reset in Bar!"
```

Here, the `Configuration` test class defined in
`dynamic_options_test.rb` by inheritance from
`Rails::Railtie::Configuration` ends up overriding the latter's
`@@options` class instance variable.

I'm not sure I understand the general use of `@@` class instance
variables in `Rails::Railstie::Configuration` (I problably wouldn't),
but this fix allows me to remove the leak without changing the
implementation.

To reproduce and confirm the fix, run this test:
```sh-session
$ bin/test test/rails_health_controller_test.rb
test/configuration/dynamic_options_test.rb --seed 15527
```

skipkayhil commented:
I'm going to close as fixed by above^
